README.md
# bayfoxr

[![Travis-CI Build Status](https://travis-ci.org/brews/bayfoxr.svg?branch=master)](https://travis-ci.org/brews/bayfoxr)
[![Coverage Status](https://coveralls.io/repos/github/brews/bayfoxr/badge.svg?branch=master)](https://coveralls.io/github/brews/bayfoxr?branch=master)

A Bayesian, global planktic foraminifera core top calibration to sea-surface temperature (SST), for R.

## What is bayfoxr?

bayfoxr is a suite of linear Bayesian calibration models for planktic core top foraminiferal δ18O (δ18Oc) and SSTs. These calibrations are especially useful because they capture the uncertainty in the relationship between modern SSTs and core top δ18Oc. This package is a companion to a paper currently under preparation for the journal "Paleoceanography and Paleoclimatology".

## A quick example

```R
library(bayfoxr)

data(bassriver)
```

The `bassriver` is example data that comes with the package. It is marine core samples from [John et al. (2008)](https://doi.org/10.1029/2007PA001465). The data.frame has two columns: "depth", giving down-core depth in meters, and "d18o", foraminifera (*Morozovella spp.*) calcite d18O samples (‰ VPDB). The core samples cover the [Paleocene-Eocene thermal maximum (PETM)](https://en.wikipedia.org/wiki/Paleocene%E2%80%93Eocene_Thermal_Maximum).

Let's run this data through our annual pooled calibration model to make inferences about past SST. *Morozovella spp.* is a nonexant species so, we're using modern planktic foraminifera as an analog with this pooled calibration.

```R
sst <- predict_seatemp(bassriver$d18o, d18osw = 0.0, 
                       prior_mean = 30.0, prior_std = 20.0)
```

The predict function then spits out a `prediction` object. Note that we need to specify d18O for seawater (`d18osw`), and a prior mean and standard deviation for our SST inference. See `help(predict_seatemp)` for more details, or `help(predict_d18oc)` for the reversed, "forward" model. 

The `sst` variable contains an ensemble rather than single prediction points because the calibration is a Bayesian regression model. This ensemble is in `sst[['ensemble']]`. Here we get median and 90% interval for the prediction:

```R
quantile(sst, probs = c(0.05, 0.50, 0.95))
```

We can also make a quick and dirty plot to visualize the inference:

```R
predictplot(x = bassriver$depth, y = sst, ylim = c(20, 40), 
            ylab = "SST (°C)", xlab = "Depth (m)")
```

## Citing bayfoxr in your research

Please cite our work if you use bayfoxr in your research. We have a paper currently in preparation and I'll be sure to update this section with the citation as soon as the paper is out.

To cite the code repository directly use:

*Malevich, Steven B., 2019. bayfoxr. \<https://github.com/brews/bayfoxr \>.*

Alternatively, you can cite the package in R's CRAN repository. You can see this information by running `citation("bayfoxr")` in an R session.

## Installation

### From CRAN

`bayfoxr` is hosted on the offical [CRAN repository](https://cran.r-project.org/package=bayfoxr). Install it from an R session with:

```R
install.packages("bayfoxr")
```

### From devtools

Bleeding edge and development versions of the package can be installed with [`devtools`](https://github.com/r-lib/devtools). Assuming you have `devtools` installed in R, you can install `bayfoxr` with:

```R
devtools::install_github("brews/bayfoxr")
```

## Support

Documentation is included in the code and can be viewed in R. Please file issues and requests in the [bug tracker](https://github.com/brews/bayfoxr/issues).

## Development
Want to contribute? We're following [Hadley's packaging workflow](http://r-pkgs.had.co.nz/) and [code style](http://adv-r.had.co.nz/Style.html). Please fork away and get in touch if you have a feature or bug fix.


NEWS.md
# bayfoxr v0.0.2.9000

Changes in this patch:

* Add CRAN install directions to README.

# bayfoxr v0.0.1

Initial release!


tests/testthat.R
library(testthat)
library(bayfoxr)

test_check("bayfoxr")


tests/testthat/test_predict_d18oc.R
context("Testing predict_d18oc()")


fake_draw_dispenser <- function(a, b, tau) {
    out <- function(foram=NULL, seasonal_seatemp=FALSE){
                    data.frame(alpha = a, beta = b, tau = tau)
    }
    out
}


seeded_predict <- function(...) {
    set.seed(123)
    predict_d18oc(...)[["ensemble"]]
}


fake_draw_get_pooledannual <- fake_draw_dispenser(a=c(1), b=c(2), tau=c(3))
fake_draw_get_hierseasonal <- fake_draw_dispenser(a=c(1, 2), b=c(2, 2), tau=c(3, 3))


test_that("Basic tests for predict_d18oc()", {
    expect_equal(seeded_predict(seatemp = 15.0, d18osw = 0.0, drawsfun = fake_draw_get_pooledannual), 
                 matrix(29.04857, ncol = 1, nrow = 1), tolerance=1e-5, check.attributes = FALSE)
    expect_equal(seeded_predict(seatemp = 15.0, d18osw = 0.0, drawsfun = fake_draw_get_hierseasonal), 
                 matrix(c(29.04857, 31.03947), ncol = 2, nrow = 1), tolerance=1e-5, check.attributes = FALSE)
    expect_equal(seeded_predict(seatemp = c(15.0, 0), d18osw = 0.0, drawsfun = fake_draw_get_hierseasonal), 
                 matrix(c(29.04857306, 0.03946753, 36.406125, 1.941525), ncol = 2, nrow = 2), 
                 tolerance=1e-5, check.attributes = FALSE)
    expect_equal(seeded_predict(seatemp = c(15.0, 0), d18osw = c(0.0, 1.0), drawsfun = fake_draw_get_hierseasonal), 
                 matrix(c(29.048573, 1.039468, 36.406125, 2.941525), ncol = 2, nrow = 2), 
                 tolerance=1e-5, check.attributes = FALSE)
})


tests/testthat/test_predict_seatemp.R
context("Testing predict_seatemp()")


fake_draw_dispenser <- function(a, b, tau) {
    out <- function(foram=NULL, seasonal_seatemp=FALSE){
        list(alpha = a, beta = b, tau = tau)
    }
    out
}


seeded_predict <- function(...) {
    set.seed(123)
    predict_seatemp(prior_mean = 15, prior_std = 20, ...)[["ensemble"]]
}


fake_draw_get_pooledannual <- fake_draw_dispenser(a=c(1), b=c(2), tau=c(3))
fake_draw_get_hierseasonal <- fake_draw_dispenser(a=c(1, 2), b=c(2, 2), tau=c(3, 3))


test_that("Basic tests for predict_seatemp()", {
    expect_equal(seeded_predict(d18oc = -3.75, d18osw = 0.0, drawsfun = fake_draw_get_pooledannual), 
                 matrix(-2.981926, ncol = 1, nrow = 1), tolerance=1e-5, check.attributes = FALSE)
    expect_equal(seeded_predict(d18oc = -3.75, d18osw = 0.0, drawsfun = fake_draw_get_hierseasonal), 
                 matrix(c(-2.981926, -2.98507), ncol = 2, nrow = 1), tolerance=1e-5, check.attributes = FALSE)
    expect_equal(seeded_predict(d18oc = c(-3.75, 0), d18osw = 0.0, drawsfun = fake_draw_get_hierseasonal), 
                 matrix(c(-2.9819263, -0.6233546, -0.3092564, -0.6707922), ncol = 2, nrow = 2), 
                 tolerance=1e-5, check.attributes = FALSE)
    expect_equal(seeded_predict(d18oc = c(-3.75, 0), d18osw = c(0.0, 1.0), drawsfun = fake_draw_get_hierseasonal), 
                 matrix(c(-2.981926, -1.120558, -0.3092564, -1.1679954), ncol = 2, nrow = 2), 
                 tolerance=1e-5, check.attributes = FALSE)
})


R/plot.R

#' Simple plot of \code{prediction} with intervals.
#'
#' @param y A \code{prediction} object to plot.
#' @param x Optional vector or NULL, indicating were \code{prediction} 
#'inferences fall along x-axis. Must be the same length as the inferred values 
#' in \code{y}.
#' @param probs Optional 3-member Vector of numerics indicating low, middle, and 
#'high probability intervals to plot. All must be <= 1.
#' @param poly_col Optional color for interval polygon.
#' @param ... Additional arguments passed to \code{plot}.
#'
#' @examples
#' data(bassriver)
#' 
#' # Using the "pooled annual" calibration model:
#' sst <- predict_seatemp(bassriver$d18o, d18osw=0.0, 
#'                        prior_mean=30.0, prior_std=20.0)
#' 
#' predictplot(x=bassriver$depth, y=sst, ylim=c(20, 40), 
#'             ylab="SST (°C)", xlab="Depth (m)")
#'
#' @export
predictplot <- function(y, x = NULL, probs = c(0.05, 0.50, 0.95), 
                        poly_col = grDevices::rgb(0, 0, 0, 0.1), ...) {
    if (length(probs) != 3) {
        stop("Length of `probs` must be 3")
    }
    if (max(probs) > 1) {
        stop("`probs` must all be <= 1")
    }

    probs <- sort(probs)
    quants <- quantile(y, probs=probs)
    y_low <- quants[, 1]
    y_mid <- quants[, 2]
    y_high <- quants[, 3]

    if (is.null(x)) {
        n_row <- nrow(y$ensemble)
        x <- seq(1, n_row)
    }
    
    graphics::plot(x = x, y = y_mid, type='b', ...)
    graphics::polygon(c(x, rev(x)), c(y_low, rev(y_high)), col = poly_col, 
                        border = NA)
}


#' Plot a \code{prediction} object.
#'
#' @param ... Arguments passed on to \code{predictplot}.
#'
#' @seealso \code{\link{predictplot}}
#'
#' @examples
#' data(bassriver)
#' 
#' # Using the "pooled annual" calibration model:
#' sst <- predict_seatemp(bassriver$d18o, d18osw=0.0, 
#'                        prior_mean=30.0, prior_std=20.0)
#' 
#' predictplot(x=bassriver$depth, y=sst, ylim=c(20, 40), 
#'             ylab="SST (°C)", xlab="Depth (m)")
#'
#' @export
plot.prediction <- function(...) {
    predictplot(...)
}


R/modelparams.R
#' Parse trace dataframe column names to get vector of available forams.
#'
#' @param d Data frame containing MCMC trace draws. Column names are model 
#'parameters with foram group name separated from model parameters name by "__"
#'
#' @return Character vector of available foram names.
#'
get_available_forams <- function(d) {
    out = c()
    for (cname in names(d)) {
        if (grepl("__", cname)) {
            # Second member should be species name.
            foram <- strsplit(cname, "__")[[1]][-1]
            out <- unique(c(out, foram))
        }
    }
    out
}


#' Get MCMC trace draws.
#' 
#' @param foram Optional. String or \code{NULL}. String indicating the foram 
#'species/subspecies to infer for hierarchical models. String must be one of 
#'"G. bulloides", "G. ruber white", "G. ruber pink", "G. sacculifer", 
#'"N. incompta", or "N. pachyderma sinistral". \code{NULL} indicates that a 
#'pooled model is desired.
#' @param seasonal_seatemp Optional boolean indicating whether to use the seasonal 
#'sea-surface temperature calibrations. Default is \code{FALSE}, i.e. using 
#'annual SST calibrations.
#'
#' @details Four calibration models are available: an "annual pooled" model, a 
#' "seasonal pooled" model, an "annual hierarchical" model, and a 
#' "seasonal hierarchical" model. This function uses magic to determine which 
# calibration model you want based on the function arguments. By default, a 
#'"pooled annual" model is used. Which is the simplest case with potential use 
#'for Deep Time reconstructions of nonexant foram species. Giving a valid string 
#'for \code{foram} will use a hierarchical model, which has foram-specific 
#'variability in calibration model parameters. Passing \code{TRUE} for 
#'\code{seasonal_seatemp} will use a model trained on season sea-surface 
#'temperatures. See reference paper for further details.
#'
#' @return Data frame with columns "alpha", "beta", "tau". Which are equal-length 
#'vectors of model parameter draws.
#'
#' @export
get_draws <- function(foram=NULL, seasonal_seatemp=FALSE) {
    # Parse internal `traces` list to get the draws we want

    # This is kinda clunky. Maybe have traces as their own S3 class with generic method to get draws.
    draw_name <- NA

    time_half <- "annual"
    foram_half <- "hierarchical"
    if (is.null(foram))
        foram_half <- "pooled"
    if (seasonal_seatemp)
        time_half <- "seasonal"
    draw_name <- paste(time_half, foram_half, sep = "_")
    
    # We're drawing from `traces` which is internal package data.
    out <- data.frame(alpha=NA, beta=NA, tau=NA)
    if (is.null(foram)) {  # Pooled
        out <- data.frame(alpha = traces[[draw_name]][["a"]],
                          beta = traces[[draw_name]][["b"]],
                          tau = traces[[draw_name]][["tau"]])
    } else {  # Hierarchical
        # Check if foram name exists and give helpful error if it doesn't.
        available_foram_names <- get_available_forams(traces[[draw_name]])
        if (foram %in% available_foram_names) {
            out <- data.frame(alpha = traces[[draw_name]][[paste("a", foram, sep = "__")]],
                              beta = traces[[draw_name]][[paste("b", foram, sep = "__")]],
                              tau = traces[[draw_name]][[paste("tau", foram, sep = "__")]])
        } else {
            stop(paste0("Bad `foram`: ", foram, "\nAvailable forams are: ", 
                        toString(available_foram_names)))
        }
    }
    out
}


R/predict.R
#' Constructor for S3 prediction class.
#'
#' @param ensemble A matrix (m x n) of the prediciton posteriors. Where m is the 
#'number of values inferred and n is the number of trace draws.
#'
#' @return A \code{prediction} object.
#'
#' @export
prediction <- function(ensemble) {
    out <- structure(list(), class = "prediction")
    out[["ensemble"]] <- as.matrix(ensemble)
    out
}


#' Quantiles for a \code{prediction}.
#'
#' @param x A \code{prediction} object.
#' @param ... Arguments to be passed on to \code{quantile}.
#'
#' @importFrom stats quantile
#' @export
quantile.prediction <- function(x, ...) {
    t(apply(X=x[["ensemble"]], MARGIN=1, FUN=quantile, ...))
}


#' Predict d18O of foram calcite given seawater temperature and seawater d18O.
#'
#' @param seatemp Numeric or vector of observed sea-surface temperatures (°C).
#' @param d18osw Numeric or vector of observed seawater d18O (‰ VSMOW). 
#' @param foram Optional. String or \code{NULL}. String indicating the foram 
#'species/subspecies to infer for hierarchical models. String must be one of 
#'"G. bulloides", "G. ruber", "T. sacculifer", "N. incompta", or 
#'"N. pachyderma". \code{NULL} indicates that a pooled model is desired.
#' @param seasonal_seatemp Optional boolean indicating whether to use the seasonal 
#'sea-surface temperature calibrations. Default is \code{FALSE}, i.e. using 
#'annual SST calibrations.
#' @param drawsfun Optional function used to get get model parameter draws. Must 
#'take arguments for "foram" and "seasonal_seatemp" and return a list with 
#'members "alpha", "beta", "tau". This is for debugging and testing. See 
#'\code{\link{get_draws}}.
#'
#' @details Four calibration models are available: an "annual pooled" model, a 
#' "seasonal pooled" model, an "annual hierarchical" model, and a 
#' "seasonal hierarchical" model. This function uses magic to determine which 
# calibration model you want based on the function arguments. By default, a 
#'"pooled annual" model is used. Which is the simplest case with potential use 
#'for Deep Time reconstructions of nonexant foram species. Giving a valid string 
#'for \code{foram} will use a hierarchical model, which has foram-specific 
#'variability in calibration model parameters. Passing \code{TRUE} for 
#'\code{seasonal_seatemp} will use a model trained on season sea-surface 
#'temperatures. See reference paper for further details.
#'
#' @return A \code{prediction} instance for inferred foraminiferal calcite 
#'d18O (‰ VPDB).
#'
#' @seealso \code{\link{predict_seatemp}}, \code{\link{predictplot}}
#'
#' @examples
#' # Infer d18Oc for a G. bulloides core top sample using annual hierarchical model.
#' # The true, d18Oc for this sample is -2.16 (‰ VPDB).
#' delo_ann <- predict_d18oc(seatemp=28.6, d18osw=0.48, foram="G. bulloides")
#' head(quantile(delo_ann, probs=c(0.159, 0.5, 0.841)))  # ± 1 standard deviation
#'
#' # Now using seasonal hierarchical model:
#' delo_sea <- predict_d18oc(seatemp=28.6, d18osw=0.48, foram="G. bulloides",
#'                           seasonal_seatemp = TRUE)
#' head(quantile(delo_sea, probs=c(0.159, 0.5, 0.841)))  # ± 1 standard deviation
#'
#' @export
predict_d18oc <- function(seatemp, d18osw, foram=NULL, seasonal_seatemp=FALSE, 
                          drawsfun=get_draws) {
    params <- drawsfun(foram = foram, seasonal_seatemp = seasonal_seatemp)
    alpha <- params[["alpha"]]
    beta <- params[["beta"]]
    tau <- params[["tau"]]

    nd <- length(seatemp)
    n_draws = length(tau)

    # Unit adjustment
    d18osw_adj <- d18osw - 0.27

    # TODO(brews): Vectorize the loop below.
    y <- matrix(nrow = length(seatemp), ncol = n_draws)
    for (i in seq(n_draws)) {
        y[, i] = stats::rnorm(n = nd, 
                              mean = alpha[i] + seatemp * beta[i] + d18osw_adj, 
                              sd = tau[i])
    }

    prediction(ensemble = y)
}


#' Predict sea-surface temperature given d18O of foram calcite and seawater d18O.
#'
#' @param d18oc Numeric or vector of observed foram calcite d18O (‰ VPDB).
#' @param d18osw Numeric or vector of observed seawater d18O (‰ VSMOW).
#' @param prior_mean Numeric indicating prior mean for sea-surface temperature (°C).
#' @param prior_std Numeric indicating prior standard deviation for sea-surface 
#'temperature (°C).
#' @param foram Optional. String or \code{NULL}. String indicating the foram 
#'species/subspecies to infer for hierarchical models. String must be one of 
#'"G. bulloides", "G. ruber", "T. sacculifer", "N. incompta", or 
#'"N. pachyderma". \code{NULL} indicates that a pooled model is desired.
#' @param seasonal_seatemp Optional boolean indicating whether to use the seasonal 
#'sea-surface temperature calibrations. Default is \code{FALSE}, i.e. using 
#'annual SST calibrations.
#' @param drawsfun Optional function used to get get model parameter draws. Must 
#'take arguments for "foram" and "seasonal_seatemp" and return a list with 
#'members "alpha", "beta", "tau". This is for debugging and testing.
#'
#' @details Four calibration models are available: an "annual pooled" model, a 
#' "seasonal pooled" model, an "annual hierarchical" model, and a 
#' "seasonal hierarchical" model. This function uses magic to determine which 
# calibration model you want based on the function arguments. By default, a 
#'"pooled annual" model is used. Which is the simplest case with potential use 
#'for Deep Time reconstructions of nonexant foram species. Giving a valid string 
#'for \code{foram} will use a hierarchical model, which has foram-specific 
#'variability in calibration model parameters. Passing \code{TRUE} for 
#'\code{seasonal_seatemp} will use a model trained on season sea-surface 
#'temperatures. See reference paper for further details.
#'
#' @return A \code{prediction} instance for inferred sea-surface temperature (°C).
#'
#' @seealso \code{\link{predict_d18oc}}
#'
#' @examples
#' data(bassriver)
#' 
#' # Using the "pooled annual" calibration model:
#' sst <- predict_seatemp(bassriver$d18o, d18osw=0.0, 
#'                        prior_mean=30.0, prior_std=20.0)
#' head(quantile(sst))  # Show only the top few values
#' 
#' predictplot(x=bassriver$depth, y=sst, ylim=c(20, 40), 
#'             ylab="SST (°C)", xlab="Depth (m)")
#'
#' @export
predict_seatemp <- function(d18oc, d18osw, prior_mean, prior_std, foram=NULL, 
                            seasonal_seatemp=FALSE, drawsfun=get_draws) {
    params <- drawsfun(foram = foram, seasonal_seatemp = seasonal_seatemp)
    alpha <- params[["alpha"]]
    beta <- params[["beta"]]
    tau <- params[["tau"]]

    nd <- length(d18oc)
    n_draws <- length(tau)

    # Unit adjustment
    d18osw_adj <- d18osw - 0.27

    # Prior mean and inverse covariance matrix
    pmu <- matrix(prior_mean, nrow = nd)
    pinv_cov <- diag(prior_std ^ (-2), nrow = nd)

    # TODO(brews): Vectorize the loop below.
    y <- matrix(nrow = length(d18oc), ncol = n_draws)
    for (i in seq(n_draws)) {
        y[, i] <- target_timeseries_pred(d18osw_now=d18osw_adj, 
                                         alpha_now=alpha[i],
                                         beta_now=beta[i], tau_now=tau[i],
                                         proxy_ts=d18oc, prior_mu=pmu,
                                         prior_inv_cov=pinv_cov)
    }

    prediction(ensemble = y)
}


#' Internal function for `predict_seatemp()`.
#'
#' @param d18osw_now Numeric or vector giving seawater d18O. Note, should be in 
#' units (‰ VPDB).
#' @param alpha_now Numeric, alpha model parameter.
#' @param beta_now Numeric, beta model parameter.
#' @param tau_now Numeric, tau model parameter.
#' @param proxy_ts Numeric or vector of proxy time series (foram d18O).
#' @param prior_mu Matrix (n X 1) giving prior mean.
#' @param prior_inv_cov Matrix (n X x) giving prior inverse covariance matrix.
#'
#' @return Sample of time time series vector conditional on the other args
#'
target_timeseries_pred <- function(d18osw_now, alpha_now, beta_now, tau_now, 
                                   proxy_ts, prior_mu, prior_inv_cov) {
    n_ts = length(proxy_ts)

    # Inverse posterior covariance matrix
    precision <- tau_now ^ (-2)
    post_inv_cov <- prior_inv_cov + precision * beta_now ^ 2 * diag(1, n_ts)

    post_cov <- solve(post_inv_cov)   # Inverse.

    # Get first factor for the mean
    mean_first_factor <- (prior_inv_cov %*% prior_mu + precision * beta_now 
                          * (proxy_ts - alpha_now - d18osw_now))
    mean_full <- t(mean_first_factor) %*% post_cov

    stats::rnorm(n = n_ts, mean = mean_full, sd = sqrt(diag(post_cov)))
}
